SlSplitPanel `snap` improvements.

Aurailus opened the pull request:
Continuing on from #1639, this PR adds new functionality to `SlSplitPanel` to allow it to have more flexible snapping logic, which allows it to function in more domains. For example, use-cases where the panel must always snap to specific percentage intervals, or where the panel must snap to the start or end of its region, but not in-between.

This is my first PR to this project. I tried to follow the existing code style as closely as possible, but please let me know if there's anything that could be improved for this PR to be merged.

## Changes

- Extend `sl-split-panel`'s snap property to accept a string in the form of `repeat(50px)` or `repeat(10%)`, which creates uniformly distributed snap points. e.g. a 450px wide split panel with `snap` set to `repeat(100px)` would generate snap points at `100px 200px 300px 400px`.
- Allow `SlSplitPanel.snap` to be set to a function to fully customize snap behaviour, allowing for logic such as `panel.snap = ({ pos, size }) => (pos < size / 2) ? 100 : (size - 100)` to define a split panel which snaps 100px from the left, and 100px from the right.
- Tweak the default snapping algorithm to prioritize the closest snap point instead of the first when there are multiple valid points within the snap threshold, allowing `snapThreshold = Infinity` to behave as you would intuitively expect it to and snap to the closest point instead of always the first.
- Update documentation.

Aurailus committed:
(sl-split-panel) Add repeat expression and function expression to [snap].

Aurailus committed:
(sl-split-panel) Improve documentation.

SabineWren commented:
Any reason why this is sitting in limbo?

claviska commented:
@Aurailus sorry, I missed [your comment](https://github.com/shoelace-style/shoelace/discussions/1639#discussioncomment-9757064) from last year.

I appreciate the PR! Some feedback:

> Allow SlSplitPanel.snap to be set to a function to fully customize snap behaviour, allowing for logic...

I'm totally fine with `.snap` accepting a function, as we have that pattern in the library already.

The only part I'm not sure about is the custom microsyntax of `repeat()`. I do like the declarative approach, but I think folks will be confused considering it's a space-delimited attribute (allowing multiple values) _except_ for this one function (only allows one value). For example, it's unclear what happens if I use `snap="25px repeat(50px)"`. It looks like it won't work glancing at the code but, as a user, I would expect that it would add a snap at 25px and at every 50px.

I'd be OK with it supporting `repeat()` as one value out of many. If the code gets too messy, I'd also be OK with removing `repeat()` altogether.

> Tweak the default snapping algorithm to prioritize the closest snap point instead of the first when there are multiple valid points within the snap threshold

Nice touch! Thanks.

> utility.ts

Nit: we prefer to keep component-specific utilities in the same file (usually as a private method) or, if they have a reasonable potential for utility in other components, in `/src/utilities/`.

Aurailus commented:
Thanks Cory! I'll take a look at getting the repeat syntax to work with other space separated values, and I'll move those helpers over to be inside the component class.

Aurailus committed:
Improve split panel repeat() syntax, code cleanup.
- Move helper methods for split panel (toSnapFunction), and type definitions (SnapFunction, SnapFunctionParams) to split-panel.component.ts
- Allow repeat() to exist within other snap points in the split panel snap property. e.g. "repeat(100px) 50% 70% 90px" is now valid.

Aurailus commented:
@claviska I've pushed a commit making the changes you requested! :) Allowing `repeat()` to be used alongside the rest of the fixed snap points actually resulted in the code being a little bit simpler overall. Thanks again!

claviska commented on docs/pages/components/split-panel.md:
````diff
@@ -239,16 +241,102 @@ To snap panels at specific positions while dragging, add the `snap` attribute wi
+
+You can also implement a custom snap function which controls the snapping manually. To do this, you need to acquire a reference to the element in Javascript and set the `snap` property. For example, if you want to snap the divider to either `100px` from the left or `100px` from the right, you can set the `snap` property to a function encoding that logic.
+
+```panel.snap = ({ pos, size }) => (pos < size / 2) ? 100 : (size - 100)```
````
```suggestion
```js
panel.snap = ({ pos, size }) => (pos < size / 2) ? 100 : (size - 100)
```
```

claviska replied:
The suggestion escaping might be broken here. The point was to put it into a proper block with JS.

claviska commented on docs/pages/components/split-panel.md:
````diff
@@ -239,16 +241,102 @@ To snap panels at specific positions while dragging, add the `snap` attribute wi
+
+```html:preview
+<div class="split-panel-snapping-repeat">
+  <sl-split-panel snap="repeat(16px) 50%">
````
This didn't match the description above.

```suggestion
  <sl-split-panel snap="repeat(100px) 50%">
```

Aurailus replied:
Sorry about that, I changed the interval after testing because it wasn't immediately apparent that the 50% was adding an extra snap point, and I forgot to actually change it in the code.

claviska commented on src/components/split-panel/split-panel.component.ts:
```diff
@@ -67,11 +86,74 @@ export default class SlSplitPanel extends ShoelaceElement {
+   * @param snap - The snap string.
+   * @returns a `SnapFunction` representing the snap string's logic.
+   */
+
```
```suggestion
```

claviska commented on docs/pages/components/split-panel.md:
````diff
@@ -239,16 +241,102 @@ To snap panels at specific positions while dragging, add the `snap` attribute wi
+</style>
+```
+
+You can also implement a custom snap function which controls the snapping manually. To do this, you need to acquire a reference to the element in Javascript and set the `snap` property. For example, if you want to snap the divider to either `100px` from the left or `100px` from the right, you can set the `snap` property to a function encoding that logic.
````
```suggestion
### Using a Custom Snap Function

You can also implement a custom snap function which controls the snapping manually. To do this, you need to acquire a reference to the element in Javascript and set the `snap` property. For example, if you want to snap the divider to either `100px` from the left or `100px` from the right, you can set the `snap` property to a function encoding that logic.
```

claviska reviewed (changes requested):
~~https://shoelace-git-fork-muniworth-aurailus-split-9fa698-font-awesome.vercel.app/components/split-panel/#snapping~~

~~The second example in this section seems to be repeating a lot more than every 100px and 50%. It's more like 10-15px.~~ — fixed in a suggestion.

Aurailus committed:
Apply suggestions from code review

Co-authored-by: Cory LaViska <cory@abeautifulsite.net>
